File: go/internal/bravo/ndjson/ndjson.go
// Package ndjson aggregates TAP-14 reader events into NDJSON records
// conforming to the tap-ndjson(7) manpage (doc/tap-ndjson.7.scd).
package ndjson

import (
	"encoding/json"
	"fmt"
	"io"

	"github.com/amarbel-llc/tap/go/internal/0/diagnostic"
)

//go:generate dagnabit export

// PlanRecord announces, up front, how many top-level test points the
// producer intends to emit. It mirrors a leading TAP `1..N` plan line
// and, when present, is the first record of the document.
type PlanRecord struct {
	Type  string `json:"type"`
	Count int    `json:"count"`
}

// TestRecord is one top-level test point with its full context.
// All fields are emitted unconditionally; nullable fields use
// pointers/slices so they serialize as JSON null when absent.
type TestRecord struct {
	Type        string          `json:"type"`
	N           int             `json:"n"`
	Description string          `json:"description"`
	OK          bool            `json:"ok"`
	Directive   *DirectiveValue `json:"directive"`
	Diagnostic  map[string]any  `json:"diagnostic"`
	Output      *string         `json:"output"`
	Subtest     []TestRecord    `json:"subtest"`
	Line        int             `json:"line"`
}

// DirectiveValue is the parsed test point directive.
type DirectiveValue struct {
	Kind   string `json:"kind"`
	Reason string `json:"reason"`
}

// BailoutRecord indicates the source stream emitted `Bail out!`.
type BailoutRecord struct {
	Type    string `json:"type"`
	Message string `json:"message"`
	Line    int    `json:"line"`
}

// SummaryRecord reports aggregate counts and validity.
type SummaryRecord struct {
	Type        string              `json:"type"`
	Passed      int                 `json:"passed"`
	Failed      int                 `json:"failed"`
	Skipped     int                 `json:"skipped"`
	Todo        int                 `json:"todo"`
	Total       int                 `json:"total"`
	PlanCount   int                 `json:"plan_count"`
	Bailed      bool                `json:"bailed"`
	Valid       bool                `json:"valid"`
	Diagnostics []SummaryDiagnostic `json:"diagnostics"`
}

// SummaryDiagnostic is one entry in SummaryRecord.Diagnostics.
type SummaryDiagnostic struct {
	Line     int    `json:"line"`
	Severity string `json:"severity"`
	Rule     string `json:"rule"`
	Message  string `json:"message"`
}

// Aggregator builds NDJSON records from reader events.
//
// Call Consume for each event in order. Call Finalize once after
// the stream is exhausted (typically at io.EOF from the reader).
//
// Subtests nest to arbitrary depth. When a test point arrives at
// depth N, any pending children at depth N+1 are embedded as that
// record's Subtest array. Records at depth > 0 are themselves
// buffered as pending children of their eventual parent at depth-1.
type Aggregator struct {
	records   []TestRecord
	planCount int
	// planLeading is set when a top-level plan line arrived before any
	// top-level test point, i.e. the producer announced its plan up
	// front. Only then is a leading `plan` record emitted.
	planLeading bool
	bailed      bool
	bailout     *BailoutRecord
	// Accumulating Output Block body for the next top-level test point.
	pendingOutput *string
	// Children buffered at their own depth, keyed by that depth. The
	// parent at depth-1 picks them up when it arrives. Initialized
	// lazily on first append.
	pendingChildren map[int][]TestRecord
}

// Output is the result of finalizing an aggregator.
type Output struct {
	Plan    *PlanRecord
	Records []TestRecord
	Bailout *BailoutRecord
	Summary SummaryRecord
}

// NewAggregator constructs an empty Aggregator ready to consume events.
func NewAggregator() *Aggregator {
	return &Aggregator{
		pendingChildren: map[int][]TestRecord{},
	}
}

// Consume feeds a single reader event into the aggregator.
func (a *Aggregator) Consume(ev diagnostic.Event) {
	switch ev.Type {
	case diagnostic.EventPlan:
		if ev.Depth == 0 && ev.Plan != nil {
			a.planCount = ev.Plan.Count
			a.planLeading = len(a.records) == 0
		}
	case diagnostic.EventTestPoint:
		if ev.TestPoint == nil {
			return
		}
		rec := buildTestRecord(ev)
		if kids := a.pendingChildren[ev.Depth+1]; len(kids) > 0 {
			rec.Subtest = kids
			delete(a.pendingChildren, ev.Depth+1)
		}
		if ev.Depth == 0 {
			if a.pendingOutput != nil {
				rec.Output = a.pendingOutput
				a.pendingOutput = nil
			}
			a.records = append(a.records, rec)
		} else {
			a.pendingChildren[ev.Depth] = append(a.pendingChildren[ev.Depth], rec)
		}
	case diagnostic.EventYAMLDiagnostic:
		if ev.Depth == 0 && len(a.records) > 0 {
			a.records[len(a.records)-1].Diagnostic = copyMap(ev.YAML)
		}
	case diagnostic.EventOutputHeader:
		if ev.Depth == 0 {
			empty := ""
			a.pendingOutput = &empty
		}
	case diagnostic.EventOutputLine:
		if ev.Depth == 0 && a.pendingOutput != nil {
			*a.pendingOutput += ev.OutputLine + "\n"
		}
	case diagnostic.EventBailOut:
		a.bailed = true
		msg := ""
		if ev.BailOut != nil {
			msg = ev.BailOut.Reason
		}
		a.bailout = &BailoutRecord{Type: "bailout", Message: msg, Line: ev.Line}
	}
}

func copyMap(m map[string]any) map[string]any {
	if m == nil {
		return nil
	}
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

// Finalize computes the aggregate summary and returns the full Output.
// readerDiags and readerSummary come from reader.Reader's Diagnostics()
// and Summary() methods; both may be nil for synthetic streams.
func (a *Aggregator) Finalize(readerDiags []diagnostic.Diagnostic, readerSummary *diagnostic.Summary) Output {
	summary := SummaryRecord{
		Type:        "summary",
		PlanCount:   a.planCount,
		Diagnostics: []SummaryDiagnostic{},
	}

	for _, r := range a.records {
		summary.Total++
		switch {
		case r.Directive != nil && r.Directive.Kind == "skip":
			summary.Skipped++
		case r.Directive != nil && r.Directive.Kind == "todo":
			summary.Todo++
		case r.OK:
			summary.Passed++
		default:
			summary.Failed++
		}
	}

	for _, d := range readerDiags {
		summary.Diagnostics = append(summary.Diagnostics, SummaryDiagnostic{
			Line:     d.Line,
			Severity: d.Severity.String(),
			Rule:     d.Rule,
			Message:  d.Message,
		})
	}

	// Orphan subtest children: any depth>0 records still in
	// pendingChildren at finalization were never claimed by a parent
	// test point. Surface them as parse diagnostics so agents know
	// the stream was malformed; the orphans themselves are dropped
	// from the output rather than synthesized into placeholder
	// records.
	for depth, kids := range a.pendingChildren {
		if len(kids) == 0 {
			continue
		}
		summary.Diagnostics = append(summary.Diagnostics, SummaryDiagnostic{
			Line:     kids[0].Line,
			Severity: "error",
			Rule:     "orphan-subtest-children",
			Message:  fmt.Sprintf("%d subtest child record(s) at depth %d had no parent test point", len(kids), depth),
		})
	}

	summary.Bailed = a.bailed
	// `valid` reports structural sanity: any error-severity diagnostic
	// (from the reader OR from our own orphan detection) drives it
	// false. It is independent of `bailed`; see RFC 0001 §"Summary
	// Record".
	summary.Valid = true
	for _, d := range summary.Diagnostics {
		if d.Severity == "error" {
			summary.Valid = false
			break
		}
	}

	var plan *PlanRecord
	if a.planLeading {
		plan = &PlanRecord{Type: "plan", Count: a.planCount}
	}

	return Output{
		Plan:    plan,
		Records: a.records,
		Bailout: a.bailout,
		Summary: summary,
	}
}

func buildTestRecord(ev diagnostic.Event) TestRecord {
	tp := ev.TestPoint
	rec := TestRecord{
		Type:        "test",
		N:           tp.Number,
		Description: tp.Description,
		OK:          tp.OK,
		Line:        ev.Line,
	}
	switch tp.Directive {
	case diagnostic.DirectiveSkip:
		rec.Directive = &DirectiveValue{Kind: "skip", Reason: tp.Reason}
	case diagnostic.DirectiveTodo:
		rec.Directive = &DirectiveValue{Kind: "todo", Reason: tp.Reason}
	}
	return rec
}

// WriteAll emits every record + bailout (if any) + summary to w as
// newline-delimited JSON.
func WriteAll(w io.Writer, out Output) error {
	enc := json.NewEncoder(w)
	enc.SetEscapeHTML(false)
	if out.Plan != nil {
		if err := enc.Encode(out.Plan); err != nil {
			return err
		}
	}
	for _, rec := range out.Records {
		if err := enc.Encode(rec); err != nil {
			return err
		}
	}
	if out.Bailout != nil {
		if err := enc.Encode(out.Bailout); err != nil {
			return err
		}
	}
	return enc.Encode(out.Summary)
}

// WriteSplit emits failures (and any bailout + summary) to failOut,
// and passes (+ bailout + summary) to passOut. passOut may be nil to
// discard passing records.
func WriteSplit(failOut, passOut io.Writer, out Output) error {
	failEnc := json.NewEncoder(failOut)
	failEnc.SetEscapeHTML(false)

	var passEnc *json.Encoder
	if passOut != nil {
		passEnc = json.NewEncoder(passOut)
		passEnc.SetEscapeHTML(false)
	}

	// The plan is a document-level record (like bailout and summary):
	// emit it first to both streams when present.
	if out.Plan != nil {
		if err := failEnc.Encode(out.Plan); err != nil {
			return err
		}
		if passEnc != nil {
			if err := passEnc.Encode(out.Plan); err != nil {
				return err
			}
		}
	}

	for _, rec := range out.Records {
		// Failure stream gets only genuine failures: !ok && directive == nil.
		// SKIP and TODO directives route to the pass stream regardless of ok.
		if !rec.OK && rec.Directive == nil {
			if err := failEnc.Encode(rec); err != nil {
				return err
			}
		} else {
			if passEnc != nil {
				if err := passEnc.Encode(rec); err != nil {
					return err
				}
			}
		}
	}

	if out.Bailout != nil {
		if err := failEnc.Encode(out.Bailout); err != nil {
			return err
		}
		if passEnc != nil {
			if err := passEnc.Encode(out.Bailout); err != nil {
				return err
			}
		}
	}

	if err := failEnc.Encode(out.Summary); err != nil {
		return err
	}
	if passEnc != nil {
		if err := passEnc.Encode(out.Summary); err != nil {
			return err
		}
	}
	return nil
}

File: rust/src/ndjson.rs
//! tap-ndjson(7) direct producer.
//!
//! Emits one spec-conforming NDJSON record per call, for producers that
//! generate test results directly rather than transforming a source TAP
//! stream. Direct producers have no source line numbers, so test and
//! bailout records carry `line: 0` ("no source line"), per the
//! tap-ndjson(7) clarification in doc/tap-ndjson.7.scd.

use std::io::{self, Write};

use serde::Serialize;
use serde_json::{Map, Value};

// Field order in these structs is the spec's SHOULD order; serde
// serializes in declaration order.

#[derive(Serialize)]
struct PlanRecord {
    #[serde(rename = "type")]
    type_: &'static str,
    count: usize,
}

#[derive(Serialize)]
struct Directive {
    kind: &'static str,
    reason: String,
}

#[derive(Serialize)]
struct TestRecord {
    #[serde(rename = "type")]
    type_: &'static str,
    n: usize,
    description: String,
    ok: bool,
    directive: Option<Directive>,
    diagnostic: Option<Map<String, Value>>,
    output: Option<String>,
    subtest: Option<Vec<TestRecord>>,
    line: usize,
}

#[derive(Serialize)]
struct BailoutRecord {
    #[serde(rename = "type")]
    type_: &'static str,
    message: String,
    line: usize,
}

#[derive(Serialize)]
struct SummaryRecord {
    #[serde(rename = "type")]
    type_: &'static str,
    passed: usize,
    failed: usize,
    skipped: usize,
    todo: usize,
    total: usize,
    plan_count: usize,
    bailed: bool,
    valid: bool,
    diagnostics: Vec<Value>,
}

fn invalid_input(msg: &str) -> io::Error {
    io::Error::new(io::ErrorKind::InvalidInput, msg.to_string())
}

/// Empty slices map to `diagnostic: null`, mirroring TapWriter's
/// not_ok_diag, which emits no YAML block for an empty slice.
fn to_diag_map(diagnostics: &[(&str, Value)]) -> Option<Map<String, Value>> {
    if diagnostics.is_empty() {
        return None;
    }
    Some(
        diagnostics
            .iter()
            .map(|(k, v)| (k.to_string(), v.clone()))
            .collect(),
    )
}

pub struct NdjsonWriter<'a> {
    w: &'a mut dyn Write,
    counter: usize,
    passed: usize,
    failed: usize,
    skipped: usize,
    todo: usize,
    plan_count: usize,
    bailed: bool,
    finished: bool,
}

impl<'a> NdjsonWriter<'a> {
    pub fn new(w: &'a mut dyn Write) -> Self {
        Self {
            w,
            counter: 0,
            passed: 0,
            failed: 0,
            skipped: 0,
            todo: 0,
            plan_count: 0,
            bailed: false,
            finished: false,
        }
    }

    pub fn count(&self) -> usize {
        self.counter
    }

    pub fn has_failures(&self) -> bool {
        self.failed > 0
    }

    fn write_record<T: Serialize>(&mut self, record: &T) -> io::Result<()> {
        serde_json::to_writer(&mut *self.w, record).map_err(io::Error::from)?;
        self.w.write_all(b"\n")
    }

    fn check_open(&self) -> io::Result<()> {
        if self.finished {
            return Err(invalid_input("record after summary"));
        }
        Ok(())
    }

    fn test(
        &mut self,
        description: &str,
        ok: bool,
        directive: Option<Directive>,
        diagnostic: Option<Map<String, Value>>,
    ) -> io::Result<usize> {
        self.check_open()?;
        if self.bailed {
            return Err(invalid_input("test record after bailout"));
        }
        self.counter += 1;
        match &directive {
            Some(d) if d.kind == "skip" => self.skipped += 1,
            Some(_) => self.todo += 1,
            None if ok => self.passed += 1,
            None => self.failed += 1,
        }
        let record = TestRecord {
            type_: "test",
            n: self.counter,
            description: description.to_string(),
            ok,
            directive,
            diagnostic,
            output: None,
            subtest: None,
            line: 0,
        };
        self.write_record(&record)?;
        Ok(self.counter)
    }

    pub fn ok(&mut self, description: &str) -> io::Result<usize> {
        self.test(description, true, None, None)
    }

    pub fn not_ok(&mut self, description: &str) -> io::Result<usize> {
        self.test(description, false, None, None)
    }

    pub fn ok_diag(
        &mut self,
        description: &str,
        diagnostics: &[(&str, Value)],
    ) -> io::Result<usize> {
        self.test(description, true, None, to_diag_map(diagnostics))
    }

    pub fn not_ok_diag(
        &mut self,
        description: &str,
        diagnostics: &[(&str, Value)],
    ) -> io::Result<usize> {
        self.test(description, false, None, to_diag_map(diagnostics))
    }

    pub fn skip(&mut self, description: &str, reason: &str) -> io::Result<usize> {
        self.test(
            description,
            true,
            Some(Directive {
                kind: "skip",
                reason: reason.to_string(),
            }),
            None,
        )
    }

    pub fn todo(&mut self, description: &str, reason: &str) -> io::Result<usize> {
        self.test(
            description,
            false,
            Some(Directive {
                kind: "todo",
                reason: reason.to_string(),
            }),
            None,
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn capture(f: impl FnOnce(&mut NdjsonWriter)) -> String {
        let mut buf = Vec::new();
        let mut w = NdjsonWriter::new(&mut buf);
        f(&mut w);
        String::from_utf8(buf).unwrap()
    }

    #[test]
    fn ok_record() {
        let out = capture(|w| {
            w.ok("socket answers").unwrap();
        });
        assert_eq!(
            out,
            "{\"type\":\"test\",\"n\":1,\"description\":\"socket answers\",\"ok\":true,\"directive\":null,\"diagnostic\":null,\"output\":null,\"subtest\":null,\"line\":0}\n"
        );
    }

    #[test]
    fn not_ok_record() {
        let out = capture(|w| {
            w.not_ok("config parses").unwrap();
        });
        assert!(out.contains("\"ok\":false"));
        assert!(out.contains("\"n\":1"));
    }

    #[test]
    fn counter_and_failures() {
        let mut buf = Vec::new();
        let mut w = NdjsonWriter::new(&mut buf);
        assert_eq!(w.count(), 0);
        assert!(!w.has_failures());
        w.ok("a").unwrap();
        w.not_ok("b").unwrap();
        assert_eq!(w.count(), 2);
        assert!(w.has_failures());
    }

    #[test]
    fn skip_directive_record() {
        let out = capture(|w| {
            w.skip("network test", "requires network").unwrap();
        });
        assert!(out.contains("\"directive\":{\"kind\":\"skip\",\"reason\":\"requires network\"}"));
        assert!(out.contains("\"ok\":true"));
    }

    #[test]
    fn todo_directive_record() {
        let out = capture(|w| {
            w.todo("future work", "not implemented").unwrap();
        });
        assert!(out.contains("\"directive\":{\"kind\":\"todo\",\"reason\":\"not implemented\"}"));
        assert!(out.contains("\"ok\":false"));
    }

    #[test]
    fn diagnostic_integers_stay_integers() {
        let out = capture(|w| {
            w.ok_diag("upstream answers", &[("keys", serde_json::json!(0))])
                .unwrap();
        });
        assert!(out.contains("\"diagnostic\":{\"keys\":0}"));
    }

    #[test]
    fn not_ok_diag_record() {
        let out = capture(|w| {
            w.not_ok_diag(
                "socket answers",
                &[("message", serde_json::json!("connection refused"))],
            )
            .unwrap();
        });
        assert!(out.contains("\"ok\":false"));
        assert!(out.contains("\"diagnostic\":{\"message\":\"connection refused\"}"));
    }

    #[test]
    fn empty_diag_slice_is_null() {
        let out = capture(|w| {
            w.not_ok_diag("broken", &[]).unwrap();
        });
        assert!(out.contains("\"diagnostic\":null"));
    }
}
